### PR TITLE
feat: add support for source-base option

### LINKED ISSUE #2218: Source map sources related to temporary output path
Hi, Evan

Thanks for a wonderful library!

I have a question / something regarding the `sources` property of outputted (inline) sourcemaps. Given I have two files,

```mjs
// foo.mjs
export default function foo () {
  console.log("YOLO!");
}
```

.. and

```mjs
// index.mjs
import foo from "./foo.mjs"

foo();
```

.. and I build like the following

```mjs
esbuild.build({
  entryPoints: ["index.mjs"],
  bundle: true,
  sourcemap: "inline"
});
```

.. I get a nice `sources` property containing `["foo.mjs", "index.mjs"]`. However, this is all outputted to stdout. I'm using `esbuild` in part with Cypress, a testing utility, and preprocessors of Cypress are provided a temporary file location in which they are expected to output bundles. Hence, we build like

```mjs
esbuild.build({
  entryPoints: ["index.mjs"],
  bundle: true,
  sourcemap: "inline",
  outfile: "/tmp/tmp-1153466-5Fh3yAWuH7tu" // <- example value
});
```

At that point, the `sources` property contains

```json
[
  "../home/jonas/projects/esbuikd-sourcemap-example/foo.mjs",
  "../home/jonas/projects/esbuikd-sourcemap-example/index.mjs"
]
```

.. and it becomes impossible to use the sources and determine file locations programmatically without also knowing the build-time CWD and the temporary output directory.

Is there a way to build to a location *and* retain "nice" and library-friendly values for `sources`?

Regards

Jonas

### PR DESCRIPTION
**Pull request by Patrick-Beuks**
Add support-base as option. This can be used when your output file is not in at the base where you want to calculate your sources from

This is inline with settings from other bundlers:
https://webpack.js.org/plugins/source-map-dev-tool-plugin/ -> fileContext
https://parceljs.org/plugin-system/source-maps/ -> projectRoot
https://github.com/terser/terser -> source-map -> base

Fixes #2218

### CODE REVIEW

File: internal/bundler/linker.go [hunk 1]
```diff
@@ -5850,6 +5850,16 @@ func (c *linkerContext) generateSourceMapForChunk(
...
+    var sourceAbsDir string
+    if c.options.SourceBase != "" {
+        var succes bool
+        sourceAbsDir, succes = c.fs.Abs(c.options.SourceBase)
+        if !succes {
+            panic("Could not determine source base")
+        }
+    } else {
+        sourceAbsDir = chunkAbsDir
+    }
```
Comment (Patrick-Beuks) [inline internal/bundler/linker.go:5862]: If there is a better way to do this if statement let me know. This is the first time writing go

File: internal/bundler/linker.go [hunk 2]
```diff
@@ -5850,6 +5850,16 @@ func (c *linkerContext) generateSourceMapForChunk(
 	dataForSourceMaps []dataForSourceMap,
 	canHaveShifts bool,
 ) (pieces sourcemap.SourceMapPieces) {
+    var sourceAbsDir string
+    if c.options.SourceBase != "" {
+        var succes bool
+        sourceAbsDir, succes = c.fs.Abs(c.options.SourceBase)
+        if !succes {
+            panic("Could not determine source base")
+        }
```
Comment (Patrick-Beuks) [inline internal/bundler/linker.go:5859]: Let me know if there is a better way to handle errors

**Comment by Patrick-Beuks** [comment]
Might fix the following issues
#2401 #2283

File: internal/bundler/linker.go [hunk 3]
```diff
@@ -5850,6 +5850,16 @@ func (c *linkerContext) generateSourceMapForChunk(
 	dataForSourceMaps []dataForSourceMap,
 	canHaveShifts bool,
 ) (pieces sourcemap.SourceMapPieces) {
+    var sourceAbsDir string
+    if c.options.SourceBase != "" {
+        var succes bool
```
Comment (jessica-jorgenson) [inline internal/bundler/linker.go:5855]:
```suggestion
        var success bool
```

File: internal/bundler/linker.go [hunk 4]
```diff
@@ -5850,6 +5850,16 @@ func (c *linkerContext) generateSourceMapForChunk(
...
+        sourceAbsDir, succes = c.fs.Abs(c.options.SourceBase)
```
Comment (jessica-jorgenson) [inline internal/bundler/linker.go:5856]:
```suggestion
        sourceAbsDir, success = c.fs.Abs(c.options.SourceBase)
```

File: internal/bundler/linker.go [hunk 5]
```diff
@@ -5850,6 +5850,16 @@ func (c *linkerContext) generateSourceMapForChunk(
...
+        if !succes {
```
Comment (jessica-jorgenson) [inline internal/bundler/linker.go:5857]:
```suggestion
        if !success {
```

**Review by jessica-jorgenson — CHANGES_REQUESTED** [review]
Spelling

**Comment by Patrick-Beuks** [comment]
Due to lack of time to update PR to resolve merge conflicts and update to latest changes I am closing this PR.

If someone wants to pick this PR up and needs information how/why this works. You can contact me.

**Pull request closed without merging**